src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -903,38 +897,40 @@ private void Dispose(bool disposing)
                     ExceptionDispatchInfo.SetCurrentStackTrace(new QuicOperationAbortedException("Read was canceled")));
             }
 
+
+            // Check if we already got final event.
+            bool releaseHandles = Interlocked.Exchange(ref _state.ShutdownDone, State.ShutdownDone_Disposed) == State.ShutdownDone_NotificationReceived;
             if (releaseHandles)
             {
                 _state.Cleanup();
             }
 
```

Comment:
This together with lines 871-877 in the original are a fix for the use after free part in https://github.com/dotnet/runtime/issues/68954
Note that the removed: `_state.ShutdownState = ShutdownState.Finished;` is actually set in `_state.Cleanup();`:
https://github.com/dotnet/runtime/blob/6ca8c9bc0c4a5fc1082c690b6768ab3be8761b11/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L94